src/main/java/com/pretz/parkingmanager/service/DuesCalculationService.java

```diff
@@ -0,0 +1,33 @@
+package com.pretz.parkingmanager.service;
+
+import com.pretz.parkingmanager.calculator.DuesCalculator;
+import com.pretz.parkingmanager.calculator.currency.CurrencyConverter;
+import com.pretz.parkingmanager.domain.ParkingSession;
+import com.pretz.parkingmanager.exception.UnknownCurrencyException;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+
+import java.math.BigDecimal;
+import java.util.Map;
+import java.util.Optional;
+
+@Service
+@RequiredArgsConstructor
+public class DuesCalculationService {
+
+    private final DuesCalculator duesCalculator;
+    private final Map<String, CurrencyConverter> currencyConverters;
+
+    BigDecimal calculateDues(ParkingSession parkingSession, String currencyCode) {
+
+        CurrencyConverter currencyConverter = getCurrencyConverter(currencyCode);
+
+        return duesCalculator.calculateDues(parkingSession, currencyConverter);
+    }
+
+    private CurrencyConverter getCurrencyConverter(String currencyCode) {
+
+        String currencyKey = currencyCode + "CurrencyConverter";
```

Comment:
wow - co to za magia?

Reply:
Magia Springa. :) Wyżej jest autowire na mapie.

    private final Map<String, CurrencyConverter> currencyConverters;

Kiedy Spring robi autowire na mapie ze Stringiem jako kluczem to dodaje do niej automatycznie wszystkie beany implementujące typ wymieniony jako wartość, zaś za klucz podstawia nazwę klasy beana.

W ten sposób znając kod mogę łatwo wywołać odpowiedni konwerter dla waluty i dodanie obsługi innych walut to tylko kwestia stworzenia nowej implementacji. Tak jak było opisane w zadaniu, może to być miejsce przyszłej zmiany.

Natomiast zamiast magicznego stringa tam powinno być CurrencyConverter.class.getSimpleName(), efekt ten sam ale bardziej koszernie